Add diagnostic filtering; range diagnostics apply to control flow constructs and compound statements

dneto0 opened the pull request:
Builds on #3683, and removes the ability to apply range diagnostic filters directly to expressions.

The only new commit is the last one (for now).

Fixes #3554 
Fixes #3689

dneto0 commented:
The substantive changes from #3683 are in the definition of the range-diagnostic filters, and the grammar changes needed to be able to apply diagnostics  to various statement forms.

dneto0 commented:
Hey @ben-clayton:  The substantive differences from your proposal in #3689 are:

- Add ability to put a filter on the block statement that is the body of a contining compound statement:
    ` loop {  .... continuning @diagnostic(off,here)  { } }`
   It's a technical fix, really
- The diagnostic filter at the start of a function declaration should have an affected range that covers the parameter declaration too, not just affecting the function body.  This is not a material difference for derivative_uniformity, but might matter later on with different diagnostics.  It does mean that the following is valid because the affected ranges are different and so the two filters do not conflict:

        @diagnostic(off,derivative_uniformity) fn foo()  // ranges over whole declaration
        @diagnostic(warn,derivative_uniformity) {    }    // ranges over the body only.

ben-clayton commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+
+Applying a diagnostic filter *DF*(*AR*,*NS*,*TR*) to a diagnostic *D* has the following effect:
+* If *D*'s [=diagnostic/triggering location=] is in *AR* and *D*'s [=diagnostic/triggering rule=] is *TR*, then:
+    * If *NS* is `off`, then *D* is discarded (i.e. deleted, or forgotten).
```
nit: I personally would have expected `off` to be listed in the `The severity of a diagnostic is one of:` bullets, and that `info` could be a severity passed to `@diagnostic`. That way, the severity list directly maps to the allowed attribute enumerators, and there (might be) fewer special carve-outs for `off`. Supporting `info` in `@diagnostic` has limited use, but could be used by the developer to filter messages at the JS level (as well as colour distinction in the console).

ben-clayton commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+</div>
+
+A <dfn noexport>global diagnostic filter</dfn> is a diagnostic filter whose [=affected range=] is the whole WGSL program.
+It is a [=directive=], thus appearing before any [=module scope|module-scope=] declarations.
```
So `enable` and `diagnostic` directives can be declared in any order, or mixed, correct?
Would be nice to demonstrate this in the examples below, except the spec doesn't have any real extensions defined yet. Never mind.

dneto0 replied:
Yes, they can be intermixed.
A reader can infer that from the translation_unit and global_directive grammar rules.

ben-clayton commented on wgsl/index.bs:
```diff
@@ -9443,26 +9678,42 @@ See [[#statements]] and [[#function-calls]].
 * Most [=built-in values=], except [=built-in values/num_workgroups=] and [=built-in values/workgroup_id=]
 * [=user-defined input datum|User-defined inputs=]
 * Certain [=built-in functions=] (see [[#uniformity-function-calls]])
 
```
Not really related to this PR, but do you think it's worth calling out that [`@interpolate(flat)` inputs are treated as non-uniform](https://github.com/gpuweb/gpuweb/issues/3668#issuecomment-1341113175)? This seems to surprise most people and many have assumed this is a bug.

dneto0 replied:
I've added it to the list of editorial issues we're tracking internally.

ben-clayton reviewed (commented):
Looks great, David. Mostly nits.

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+* A [=diagnostic/severity=].
+* A [=diagnostic/triggering rule=].
+* A [=diagnostic/triggering location=].
+* Optionally, details describing how the [=diagnostic/triggering rule=] applies in this case.
```
Sure, I can remove it, as it's not essential to the rules.

The thought was that when the developer sees the diagnostic, the diagnostic should say how the rule applies. e.g. what types were instantiated, notes, blah blah blah. That's necessary for good product quality but can't really be spec'd.

ben-clayton commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+The <dfn dfn-for="diagnostic">severity</dfn> of a diagnostic is one of:
+: <dfn dfn-for="severity" noexport>error</a>
+:: The diagnostic is an error.
+     This is the greatest severity, and corresponds to a [=shader-creation error=] or to a [=pipeline-creation error=].
```
ack

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+
+Triggered diagnostics [=behavioral requirement|will=] be processed as follows:
+1. For each diagnostic *D*, find the [=diagnostic filter=]
+    with the smallest [=affected range=] that contains D's [=diagnostic/triggering location=], and which has the same [=diagnostic/triggering rule=].
```
There is a note at the end of the "Diagnostic Filtering" section saying that WGSL's diagnostic filters nest perfectly.

Saying "smallest" is the minimal thing, without needing to introduce nesting.

ben-clayton replied:
ack.

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+
+An implementation may support filtering triggering rules not specified here,
+provided they are spelled differently from the standard triggering rules.
+A [=diagnostic filter=] with an unrecognized [=diagnostic/triggering rule=] [=behavioral requirement|will=] be ignored.
```
I filed https://github.com/gpuweb/gpuweb/issues/3714

ben-clayton replied:
Perfect, thank you!

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+* *NS*: a new severity
+* *TR*: a [=diagnostic/triggering rule=]
+
+Applying a diagnostic filter *DF*(*AR*,*NS*,*TR*) to a diagnostic *D* has the following effect:
```
There's redundant machinery here between what it means to apply a diagnostic filter to a diagnostic that doesn't match, vs. finding the smallest matching filter up in the description of processing diagnostics overall.

It's more complicated than it strictly needs to be.

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+Two [=diagnostic filters=] *DF*(*AR1*,*NS1*,*TR1*) and *DF*(*AR2*,*NS2*,*TR2*) <dfn dfn-for="diagnostic">conflict</dfn> when:
+* (*AR1* &equals; *AR2*), and
+* (*TR1* &equals; *TR2*), and
+* (*NS1* &ne; *NS2*).
```
Ooh. I didn't remember the uniqueness rule. You're right. 
I would favour allowing duplicating diagnostics on the same object.

dneto0 commented on wgsl/index.bs:
```diff
@@ -9443,26 +9678,42 @@ See [[#statements]] and [[#function-calls]].
+If [=uniformity analysis=] cannot prove that a particular [[#collective-operations|collective operation]] executes in [=uniform control flow=], then
+* If the operation is a
+    [=builtin functions that compute a derivative|builtin function that computes a derivative=], then
+    a [=trigger/derivative_uniformity=] [=diagnostic=] is [=triggered=].
```
Defined a new concept of triggering a "uniformity failure", and then changed the various "rejects a program" instances to "triggers a uniformity failure"

jrprice commented on wgsl/index.bs:
```diff
@@ -9739,13 +9742,14 @@ The remaining subsections specify a static analysis that verifies that
-      - Linear time complexity (in the number of tokens in the program)
-      - Refactoring a piece of code into a function, or inlining a function, cannot make a shader invalid if it was valid before the transformation
-      - If the analysis refuses a program, it provides a straightforward chain of implications that can be used by the user agent to craft a good error message
+      - Sound, meaning that a [=uniformity failure=] [=behavioral requirement|will=] be triggered for program that would break the uniformity requirements of builtins.
```
Nit: "be triggered for <missing word> program"

dneto0 replied:
fixed.

dneto0 commented on wgsl/index.bs:
```diff
@@ -6717,14 +6935,25 @@ When an identifier is used as a [=syntax/callable=] item, it is one of:
+  <dfn for=syntax>call_phrase</dfn> :
+
+    | [=syntax/callable=] [=syntax/argument_expression_list=]
+</div>
```
That has been pulled out into #3718

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,191 @@ When unclear from context, this specification indicates
+    * If such a filter exists, apply it to *D*.  This may discard *D* or change *D*'s [=diagnostic/severity=].
+    * Otherwise *D* remains unchanged.
+2. If at least one remaining diagnostic has [=severity/error=] severity, then:
+    * Other diagnostics *may* be discarded, including other diagnostics with [=severity/error=] severity.
```
Will address this.

It seems this would also apply to diagnostics with "info" level, since a warning may have been downgraded to "info".
The feasibility of the combinatorics are the same no matter what the severity level.

dneto0 replied:
This should be addressed now.
For any particular rule, at most one info message need be generated.
For any particular rule, at most one warning message need be generated (and info messages for the same triggering rule may be discarded).

dneto0 commented:
Squashed all the commits together and rewrote the commit message.

dneto0 commented:
Rebased to pick up #3725

dneto0 commented on wgsl/index.bs:
```diff
@@ -9754,6 +9754,17 @@ There is no risk of being trapped in a cycle, as recurrence is forbidden in the
+* Create a witness set containing a single source location.
+* Form a witness set from the union of two others.
+
+Witness sets can be implemented such that their total cost across the entire uniformity analysis is linear in the source program size.
```
If you implement with a persistent trees, in the sense of https://en.wikipedia.org/wiki/Persistent_data_structure then  you get this.

dneto0 replied:
Remove this sentence per feedback form @jimblandy

alan-baker commented on wgsl/index.bs:
```diff
@@ -9586,20 +9851,26 @@ value|non-uniform=] during the execution of the function call.
 * For each [=Value_return_i=] node, record which [=param_i=] nodes are reachable from it.
-* Look at which nodes are reachable from [=RequiredToBeUniform=].
-    * If this set includes the node [=MayBeNonUniform=], then reject the program.
-    * If this set includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=].
+* Examine the [=RequiredToBeUniform=] node. Let *W* be its associated [=witness set=]. Let *R* be the set of nodes reachable from [=RequiredToBeUniform=].
+    * If *R* includes the node [=MayBeNonUniform=], then trigger a [=uniformity failure=].
+         * If the uniformity failure triggers [=diagnostics=], then the [=diagnostic/triggering locations=] are taken from *W*.
+    * If *R* includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=], with [=witness set=] *W*.
     * Otherwise, the [=call site tag=] is [=CallSiteNoRestriction=].
-    * For each [=param_i=] in this set, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniform=].
+    * For each [=param_i=] in *R*, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniform=], with [=witness set=] *W*.
```
This seems like it could generate diagnostics with incorrect triggering locations, which seems undesirable. Consider:
```
fn bar(c1 : bool, c2 : bool) -> vec4<f32> {
  var x : vec4<f32>;
  if c1 {
    x += textureSample(...);
  }
  if c2 {
    x + textureSample(...);
  }
  return x;
}

var a : f32;
fn foo() {
  let uniform = true;
  let nonuniform = a < 0.5;
  let x = bar(uniform, nonuniform);
}
```

It seems permissible that a diagnostic for a uniformity failure be issued with the triggering location of the first textureSample call despite that being provably uniform. This happens because both parameters in bar are reachable from RequiredToBeUniform, which has a witness set of both textureSample locations.

dneto0 replied:
Thanks for catching this again.

I've reworked so that triggering locations are tracked through the CallSiteRequiredToBeUniform and ParameterRequiredToBeUniform tags.   

Also, per community group feedback, the algorithm in the spec is much closer to a "good" implementation, as in Tint. The graph construction takes severities into account.  This lets us track (triggering location, rule) precisely according to the severity associated with the triggering call site.

jimblandy commented on wgsl/index.bs:
```diff
@@ -534,7 +534,196 @@ When unclear from context, this specification indicates
+  <tr>
+    <td><dfn noexport dfn-for="trigger">derivative_uniformity</dfn>
+    <td>[=severity/error=]
+    <td>A call to a [=builtin functions that compute a derivative|builtin function that computes a derivative=].
```
Since we're defining the triggering location, shouldn't this spell out exactly what the location is, in terms of the grammar?

dneto0 replied:
I've updated this to say "call site", and then refined "call site" itself to say it's the location of the first token in the parsed call_phrase.
Perhaps call site should really be the range of source locations across the entire call_phrase.  It would work either way but then we'd be talking about source ranges all the time.

jimblandy commented on wgsl/index.bs:
```diff
@@ -9444,26 +9675,44 @@ See [[#statements]] and [[#function-calls]].
+* If a uniformity failure is triggered for a
+    [=builtin functions that compute a derivative|builtin function that computes a derivative=], then
+    a [=trigger/derivative_uniformity=] [=diagnostic=] is [=triggered=].
+    The [=diagnostic/triggering location=] for the diagnostic is the [=call site=] of that builtin.
```
Is "call site" the right term to use here? The definition says:

> Call sites are a [dynamic context](http://raw.githack.com/gpuweb/gpuweb-pr-previews/e2ee9948ce503be489787d630290d5c9dbf2ae7c/gpuweb/gpuweb/3713/wgsl/index.html#dynamic-context). As such, the same textual location may represent multiple call sites.

This sounds to me like "call site" should not be used to refer to a specific textual location, but I'm not sure.

dneto0 replied:
Yes, good call.  Most times I think of call site as a text location, but you're right that it's defined that way.
Fixed this spot and also the spot at the definition of derivative_uniformity in 2.2.1 Filterable Triggering Rules.

jimblandy commented on wgsl/index.bs:
```diff
@@ -9504,12 +9754,24 @@ There is no risk of being trapped in a cycle, as recurrence is forbidden in the
+Additionally, the analysis computes and propagates locations of [=call sites=] for [=builtin functions that compute a derivative=].
+This is used to [=triggered|trigger=] and filter [=diagnostics=] for [=uniformity failures=] on those calls.
+A set of such source locations is called a <dfn noexport>witness set</dfn>.
+Witness sets only support adding elements, not removing elements.
+In particular, a witness set is only formed via the following operations:
+* Create an empty set.
+* Create a witness set containing a single source location.
+* Form a witness set from the union of two others.
+
+Witness sets can be implemented such that their total cost across the entire uniformity analysis is linear in the source program size.
```
We don't usually talk about implementation details like this in the spec. It sounds like you're trying to reassure the reader that it won't be too hard to implement, but that's not the spec's job.

dneto0 replied:
Removed this.

jimblandy commented on wgsl/index.bs:
```diff
@@ -9586,20 +9851,26 @@ value|non-uniform=] during the execution of the function call.
-    * If this set includes the node [=MayBeNonUniform=], then reject the program.
-    * If this set includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=].
+* Examine the [=RequiredToBeUniform=] node. Let *W* be its associated [=witness set=]. Let *R* be the set of nodes reachable from [=RequiredToBeUniform=].
+    * If *R* includes the node [=MayBeNonUniform=], then trigger a [=uniformity failure=].
+         * If the uniformity failure triggers [=diagnostics=], then the [=diagnostic/triggering locations=] are taken from *W*.
```
This language seems to assume that, once we've triggered the uniformity failure, we don't need to worry about what tags to assign the function and its parameters, but if the diagnostic is filtered off, then the algorithm still needs to make a judgment about what call site tag to assign the function.

If the intention is to fall through to the following rules when the diagnostic is triggered off, then that should be said more explicitly.
I guess this assigns

dneto0 replied:
I think I still need to address this.

dneto0 replied:
This whole algorithm has been rephrased for clarity.  This bit has a big "Otherwise" surrounding the call site tag and paramter tag updates.

dneto0 commented:
Just did a rebase. No content change since the last reviews. (Except for fixing up cross-references to IDL in WebGPU API spec)

dneto0 commented on wgsl/index.bs:
```diff
@@ -534,7 +534,196 @@ When unclear from context, this specification indicates
+    <tr><td>Beginning of a [=statement/loop=] statement.<td>The loop statement.
+    <tr><td>Beginning of a [=statement/while=] statement.<td>The while statement: both the condition expression and the loop body.
+    <tr><td>Beginning of a [=statement/for=] statement.<td>The for statement: the [=syntax/for_header=] and the loop body.
+    <tr><td>Immediately before the opening brace (`'{'`) of the [=loop body=] of a [=statement/loop=], [=statement/while=], or [=statement/for=] loop.<td>The [=loop body=].
```
@jrprice brought up the following example.  Suppose we add an "unknown_diagnostic" triggering rule such as for #3714.

Then this should "work", mean trigger an error.
```
fn foo() {
       @diagnostic(off,derivative_uniformity_misspelled) @diagnostic(error,unknown_diagnostic) { return; }
}
```
This would require the affected range of the second `@diagnostic` attribute extend all the way to the start of the line, i.e. to its left.

alan-baker replied:
I still prefer clarifications here. Maybe the link should be to the grammar instead of the statement section or just a general statement that beginning means according to the grammar and an example to clarify.

dneto0 replied:
You fixed this yourself. :-)

dneto0 commented:
FYI:  Still TODO is to accommodate #3677

dneto0 commented:
The build-valdate-publish failure looks like a bikeshed dependencies problem. Is it transient?

jrprice commented on wgsl/index.bs:
```diff
@@ -513,7 +514,205 @@ When unclear from context, this specification indicates
+    <td><dfn noexport dfn-for="trigger">derivative_uniformity</dfn>
+    <td>[=severity/error=]
+    <td>The [=call site=] for any of the
+        [=builtin functions that compute a derivative|builtin function that computes a derivative=].
```
Nit: this reads a little strange in the rendered version:
"for any of the builtin function that computes a derivative"
Suggest removing "of the".

jrprice commented on wgsl/index.bs:
```diff
@@ -513,7 +514,205 @@ When unclear from context, this specification indicates
+Future versions of this specification may remove a particular rule or weaken its default severity
+(i.e. replace its current default with a less severe default) and still be deemed as satisfying backward compatibility.
+For example, a future version of WGSL may change the default severity for [=trigger/derivative_uniformity=] from `error` to either `warning` or `info`.
+After such a change to the specification, previously valid programs would remain valid.
```
I don't know how much it matters, but I can image scenarios where this might not quite be true. Consider an implementation that produces a warning for an unrecognized diagnostic rule (as Tint does currently), that also provides an implementation-specific rule for controlling the severity of unrecognized diagnostic rules. A shader author might want unrecognized rules to be hard errors, so they write this:
```
diagnostic(error, unrecognized_diagnostic_rule);
diagnostic(off, derivative_uniformity);
```
Now, if we were to remove the `derivative_uniformity` rule from WGSL in the future, this shader would start producing a shader-creation error when it didn't previously.

Maybe this is a corner case that we are OK with? Or, we have to guarantee that we will never remove a rule from the spec (but certain rules may become obsolete instead).

dneto0 replied:
Sigh.  When first writing this up, I didn't think of the idea of "unrecognized_diagnostic_rule". 

You're right, we could have a an explicit list of obsoleted rules. They are always "off" in some sense to be worked out. ?

jrprice commented on wgsl/index.bs:
```diff
@@ -513,7 +514,205 @@ When unclear from context, this specification indicates
+    <tr><td>Beginning of a [=statement/loop=] statement.<td>The loop statement.
+    <tr><td>Beginning of a [=statement/while=] statement.<td>The while statement: both the condition expression and the loop body.
+    <tr><td>Beginning of a [=statement/for=] statement.<td>The for statement: the [=syntax/for_header=] and the loop body.
+    <tr><td>Immediately before the opening brace (`'{'`) of the [=loop body=] of a [=statement/loop=], [=statement/while=], or [=statement/for=] loop.<td>The [=loop body=].
```
Nit: `for` and `while` loop bodies are `compound_statements`, so should maybe be listed in the note below instead?

jrprice commented on wgsl/index.bs:
```diff
@@ -513,7 +514,205 @@ When unclear from context, this specification indicates
+</table>
+
+Note: A [=function body=], a [=case clause=], a [=default-alone clause=],
+and the bodies of [=syntax/else_if_clause=] and [=syntax/else_clause=] are themselves [=compound statements=].
```
Nit: Also the body of an if clause? (this isn't covered by any rule of the table above)

jrprice commented on wgsl/index.bs:
```diff
@@ -966,6 +1165,21 @@ The spelling of the token may be the same as an [=identifier=], but the token do
+
+A <dfn noexport>diagnostic rule name</dfn> is a [=token=] used to name a diagnostic [=diagnostic/triggering rule=],
+in either a [=range diagnostic filter=] or a [=global diagnostic filter=].
+The spelling of the token may be the same as an [=identifier=], [=keyword=], or [=reserved word=],
```
Q: What's the reason for allowing keywords and reserved words as diagnostic rule names? Tint currently does not support them here, and while it is surely implementable, I'm not sure there's value for authors.

dneto0 replied:
Same for reserved words too.

dneto0 replied:
will remove

jrprice commented on wgsl/index.bs:
```diff
@@ -9453,26 +9696,47 @@ See [[#statements]] and [[#function-calls]].
+> For example, a non-uniform control dependency arises when different invocations compute
+    different values for the condition of an
+    [=statement/if=], [=statement/break-if=], [=statement/while=], or [=statement/for=],
+    or different values for the selector of a [=statement/switch=].
```
I'm not sure if this was intended to be exhaustive, but the other case is the LHS of a short-circuiting binary operator.

dneto0 replied:
Will add it.

jrprice commented on wgsl/index.bs:
```diff
@@ -9498,13 +9762,14 @@ The remaining subsections specify a static analysis that verifies that
 
-Each function is analyzed, verifying that there is a context where it is safe to call this function. It rejects the program as invalid if there is no such context.
+Each function is analyzed, verifying that there is a context where it is safe to call this function.
+It triggers a [=uniformity failure=] if there is no such context.
```
Nit: This sounds like we are trying to determine if there exists "a context" where it is safe to call the function. We are actually trying to ensure that _all_ contexts where the function is called are safe.

dneto0 replied:
Rewrote as:

> Each function is analyzed, trying to ensure two things:

> * Uniformity requirements are satisfied when it calls other functions, and

> * Uniformity requirements are satisfied whenever it is called.

> A [uniformity failure](https://github.com/gpuweb/gpuweb/pull/3713#uniformity-failure) is triggered if either of these two checks fail.


> As part of this work, the analysis computes metadata...

jrprice commented on wgsl/index.bs:
```diff
@@ -9513,49 +9778,81 @@ There is no risk of being trapped in a cycle, as recurrence is forbidden in the
+      <td><dfn noexport>CallSiteRequiredToBeUniform.S</dfn>,<br>
+          where |S| is one of the severities: [=severity/error=], [=severity/warning=], or [=severity/info=].
+      <td>The function must only be called from [=uniform control flow=].
+          Otherwise a diagnostic of with severity |S| will be triggered.
```
Nit: "of with"

jrprice commented on wgsl/index.bs:
```diff
@@ -9576,8 +9873,13 @@ value|non-uniform=] during the execution of the function call.
+      <td><dfn noexport>ParameterRequiredToBeUniform.S</dfn>,<br>
+          where |S| is one of the severities: [=severity/error=], [=severity/warning=], or [=severity/info=].
+      <td>The parameter must be a [=uniform value=].
+          Otherwise a diagnostic of with severity |S| will be triggered.
```
Nit: "of with"

jrprice commented on wgsl/index.bs:
```diff
@@ -10040,16 +10367,31 @@ Note: Refer to [[#func-var-value-analysis]] for the definition of *Vout*(*call*)
+- [[#arrayLength-builtin|arrayLength]]:
+    - Has a [=function tag=] of [=NoRestriction=].
+    - Has a [=call site tag=] of [=CallSiteNoRestriction=].
+    - The input parameter `p` has a [=parameter tag=] of [=ParameterNoRestriction=]
```
I don't think the special case for `arrrayLength` is needed at all any more. It's fine for its parameter tag to be `ParameterRequiredToBeUniformForReturnValue` as per the defaults above (this is what Tint does) and this is future proof for a world where we have variable pointers.

jrprice commented on wgsl/index.bs:
```diff
@@ -9597,29 +9899,50 @@ value|non-uniform=] during the execution of the function call.
+    * Let *R.S* be the set of nodes reachable from [=RequiredToBeUniform.S=].
+    * Let *W* be the [=witness set=] associated with [=RequiredToBeUniform.S=].
+    * If *R.S* includes the node [=MayBeNonUniform=], then trigger a [=uniformity failure=]:
+         * For each [=witness=] *w* in *W*, [=triggered|trigger=] a diagnostic with severity *S*, using the call site and triggering rule from *w*.
```
I'm trying to understand why we need witness sets at all, now that we explicitly encode the diagnostic severity in the graph. I think this line is the only place where they have an effect, and that effect is to propagate the triggering rule to the diagnostic that is raised (i.e. `derivative_uniformity` vs unnamed).

The triggering rule is used to determine the diagnostic severity, but we've already figured that out at the callsite and propagated it through the graph. So the witness set machinery feels redundant to me, and it adds additional complexity to the analysis that does not reflect the implementation.

Please let me know if I'm missing something here.

dneto0 replied:
>  I think this line is the only place where they have an effect,

Yes.

> and that effect is to propagate the triggering rule to the diagnostic that is raised (i.e. derivative_uniformity vs unnamed).

Not quite.  We have to create diagnostics, and a triggering location is one of the 3 defining properties of a diagnostic.  So we needed the source location, and that has to be propagated up.  "Witness" = (source location, severity).

Funny thing.  I had deleted the witness machinery, and for a while I had "TODO"s to track locations, then added them back exactly for this reason.

alan-baker replied:
The witness set exists to propagate the triggering location out of a function. It still seems required to me.

alan-baker commented on wgsl/index.bs:
```diff
@@ -9597,29 +9899,50 @@ value|non-uniform=] during the execution of the function call.
+    * The [=call site tag=] is initialized to [=CallSiteNoRestriction=].
+    * The [=parameter tag=] for each [=param_i=] is initialized to [=ParameterNoRestriction=].
+    * The [=pointer parameter tag=] for each [=param_i=], if it exists, is initialized to [=PointerParameterNoRestriction=].
+* For each severity *S* in the order {[=severity/error=], [=severity/warning=], [=severity/info=]}, perform the following:
```
Previously the analysis removed nodes after traversing from RequiredToBeUniform. That has a knock-on effect with how the rules for function calls are constructed later. Consider the following:

```
fn foo(c : bool) -> bool {
  @diagnostic(info, derivative_uniformity)
  if (c) {
    let x = textureSample(...);
  }
  return c;
}

var<private> x : bool;
fn bar() {
  let tmp = foo(x);
  if (tmp) {
    let s = textureSample(...);
  }
}
```

`foo` gets tagged with `RequiredToBeUniform.info` and parameter `c` gets tagged with `ParameterRequiredToBeUniform.info`. Therefore, in `bar`, the graph is disconnected between before and after the function call to `foo`. 

So there will be an info diagnostic triggered for the sample in `foo`, but it will miss the error diagnostic for the sample in `bar. The easiest change might be to always connect `CF_after` to the last `CF_i`.

dneto0 replied:
Yeah, there were two subtle things:
- statefulenss: removing nodes as you go along
- and the early exit implied by the "reject the shader", which now can't just be "reject the shader".

jrprice replied:
The early exit change isn't enough here. Right now (with this PR), the `ParameterRequiredToBeUniform.S` and `ParameterRequiredToBeUniformForReturnValue` states are mutually exclusive, which is why this doesn't work.

In Tint, we split the tag to allow those two states to co-exist ([see here](https://source.chromium.org/chromium/chromium/src/+/627b57a43f94511a0cfdea3aab071282cc2bb64b:third_party/dawn/src/tint/resolver/uniformity.cc;l=146-149)). This also ties in to our solution for #3677, which has a similar problem.

alan-baker commented on wgsl/index.bs:
```diff
@@ -10040,16 +10367,31 @@ Note: Refer to [[#func-var-value-analysis]] for the definition of *Vout*(*call*)
+- A call to a function in
+    [[#derivative-builtin-functions]], [[#texturesample]], [[#texturesamplebias]], and [[#texturesamplecompare]]:
+    - Has a [=function tag=] of [=ReturnValueMayBeNonUniform=].
+    - Has a [=call site tag=] of [=CallSiteRequiredToBeUniform.S=]:
+        - Where *S* is:
+            - the NS (new severity) parameter of the [=nearest enclosing diagnostic filter=]
+                for the combination of the call site and the [=trigger/derivative_uniformity=] triggering rule.
```
`NS` might be `off`, but above you only created nodes for error, warning, and info. It should probably be called out that `off` is not included in the graph.

jrprice commented on wgsl/index.bs:
```diff
@@ -10418,11 +10777,11 @@ end of the loop body (CF_after_if).
 uses a user-defined function call.
-The analysis tags both parameters of `scale` as
-[=ParameterRequiredToBeUniformForReturnValue=].
+The analysis tags the [=parameter return tag=] both parameters of `scale` as
```
Nit: I think there's a missing "of" or something here.
"tags the parameter return tag [of?] both parameters "

jrprice reviewed (approved):
LGTM modulo the outstanding minor comments. I'm comfortable that this closely matches what we have implemented in Tint.

dneto0 commented on wgsl/index.bs:
```diff
@@ -9597,29 +9914,53 @@ value|non-uniform=] during the execution of the function call.
+    * For each formal parameter that is a pointer in the [=address spaces/function=] address space,
         create a <dfn noexport>Value_return_i</dfn> node.
+        This represents the function's effect on the uniformity of the value pointed to by the *i*<sup>th</sup> parameter.
 * Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next sections ([[#func-var-value-analysis]], [[#uniformity-statements]], [[#uniformity-function-calls]], [[#uniformity-expressions]]), using [=CF_start=] as the starting control-flow for the function's body.
```
Add a sub-bullet saying.  

> The nodes added in this step are called `<dfn noexport>`interior nodes`</dfn>`

dneto0 commented on wgsl/index.bs:
```diff
@@ -9597,29 +9914,53 @@ value|non-uniform=] during the execution of the function call.
+    * The [=pointer parameter tag=] for each [=param_i=], if it exists, is initialized to [=PointerParameterNoRestriction=].
+* For each severity *S* in the order {[=severity/error=], [=severity/warning=], [=severity/info=]}, perform the following:
+    * Let *R.S* be the set of unvisited nodes reachable from [=RequiredToBeUniform.S=].
+    * Mark the nodes in *R.S* as visited.
```
> Mark the [=interior nodes=] in *R.S* as visited.

The reason we want this is so that we don't think of things like param_i as having been visited.  We still need to be able to "reach" them with a search via unvisited nodes.

dneto0 commented on wgsl/index.bs:
```diff
@@ -9597,29 +9914,53 @@ value|non-uniform=] during the execution of the function call.
+             set the [=call site tag=] to [=CallSiteRequiredToBeUniform.S=], and set its [=witness set=] to *W*.
+        * For each [=param_i=] in *R.S*, if its corresponding [=parameter tag=] has not been updated since initialization, then
+             set that tag to [=ParameterRequiredToBeUniform.S=], and set its [=witness set=] to *W*.
+* Mark all the nodes as unvisited.
```
> Mark all the [=interior nodes=] as unvisited.

dneto0 commented on wgsl/index.bs:
```diff
@@ -9932,10 +9944,11 @@ The following algorithm describes how to compute these tags for a given function
     * If *VR* includes [=MayBeNonUniform=], then set the [=function tag=] to [=ReturnValueMayBeNonUniform=].
-    * For each [=param_i=] in *VR*, if the corresponding [=parameter tag=] has not been updated since initialization,
-         then set it to [=ParameterRequiredToBeUniformForReturnValue=].
+    * For each [=param_i=] in *VR*, if the corresponding [=parameter return tag=] has not been updated since initialization,
```
We don't need the condition that it has not been updated since initialization.
It's ok to set it multiple times because there is only one bad value.

I only needed the condition when this was updating the parameter tag that needed to keep the severity information (and witness sets).

dneto0 commented on wgsl/index.bs:
```diff
@@ -10777,7 +10777,7 @@ end of the loop body (CF_after_if).
 This example is modification of the [[#uniformity-example1|first example]], but
 uses a user-defined function call.
-The analysis tags the [=parameter return tag=] both parameters of `scale` as
+The analysis tags the [=parameter return tag=] of both parameters of `scale` as
```
Recommend replacing "tags" with "sets", to avoid confusion: 

> The analysis sets the ....

dneto0 reviewed (commented):
I think there is a substantive clarification that is needed, i.e. the "interior nodes" thing so that visitation semantics are handled correctly.

dneto0 reviewed (commented):
LGTM to me now!

I can't "approve" the PR.

dneto0 commented:
This is ready for review @jimblandy @litherum

dneto0 commented:
Rebased and resolved a trivial conflict.

dneto0 commented:
Rebased and resolved conflict.

dneto0 commented:
Rebased and resolved conflicts

dneto0 commented:
Rebased. No conflicts.

dneto0 commented:
Simplified "witness set" to "potential-trigger-set".  

   
    Before, a witness set is a set of  pairs of (call site, triggering rule)
    Now, a potential-trigger-set is set of triggering rules.
    There are only two possible members:
      - derivative_uniformity
      - an unnamed triggering rule, for unfilterable uniformity requirements
         for compute shader barrier-like synchronization




Rebased, and resolved cosmetic conflict w.r.t. marker spans.

jimblandy reviewed (approved):
This looks good to me. I do have some requested changes, but if they're controversial I'm fine with merging this and discussing them in follow-up issues/PRs.

dneto0 commented on wgsl/index.bs:
```diff
@@ -559,7 +560,207 @@ When unclear from context, this specification indicates
+
+An implementation may support filtering triggering rules not specified here,
+provided they are spelled differently from the standard triggering rules.
+A [=diagnostic filter=] with an unrecognized [=diagnostic/triggering rule=] [=behavioral requirement|will=] be ignored.
```
For my own context, 'clippy' is an actual thing in Rust culture

https://doc.rust-lang.org/nightly/clippy/

dneto0 replied:
Will file a followup issue to discuss this suggestion.

dneto0 replied:
See https://github.com/gpuweb/gpuweb/issues/3927

dneto0 commented on wgsl/index.bs:
```diff
@@ -9505,29 +9822,53 @@ value|non-uniform=] during the execution of the function call.
+             set the [=call site tag=] to [=CallSiteRequiredToBeUniform.S=], and set its [=witness set=] to *W*.
+        * For each [=param_i=] in *R.S*, if its corresponding [=parameter tag=] has not been updated since initialization, then
+             set that tag to [=ParameterRequiredToBeUniform.S=], and set its [=witness set=] to *W*.
+* Mark all the [=interior nodes=] as unvisited.
```
I think it is actually necessary.  The same node might be used to generate a diagnostic due to a call in this function, and also needs to be analyzed for its nonuniform effects on the return value.

Consider this program:
```
@group(0) @binding(0) var t: texture_2d<f32>;
@group(0) @binding(1) var s: sampler;
fn foo(x: vec2<f32>) -> vec2<f32> {
  let c2 = textureSample(t,s,x).xy;
  if (c2.x < 0.0) {
     _ = textureSample(t,s,c2);
  } else {
     return vec2f();
  }
  return c2;
}

@fragment
fn main(@builtin(position) p: vec4<f32>) -> @location(0) vec4<f32> {
  return vec4(foo(p.xy),vec2f());
}
```

The gist of the issue is that the second call to textureSample may be in nonuniform control flow because of the non-uniformity of the c2 value.    
But the non-uniformity of the result also depends on the non-uniformity of the c2 value, because different values are returned from the two different paths out of the "if".

I'm attaching the analysis graph for the foo function.
![tricky](https://user-images.githubusercontent.com/13054574/223232747-85a104a9-f305-4d75-b7a0-a2de82c03f3d.svg)

The second call to textureSample generates an error because there's a path from foo.RequiredToBeUnifom to foo.MayBeNonUniform. Furthermore that path goes through foo.textureSample_return_value, which represents the uniformity summary for the return value of the *first* textureSample call.   In fact that's the only path from RequiredToBeUniform to MayBeNonUniform.  So after generating the diagnostic, the *interior* node foo.textureSample_return_value is marked as unvisited.

Now, observe that *all* paths from `foo.Value_return` to `foo.param_x` actually go through `foo.textureSample_return_value`.  So if we *didn't mark that interior node as unvisited, then we'd miss the fact that the uniformity of the return value depends on the return value of the 'x' parameter.
